### DISCUSSION TITLE
sqalchemy core  batch insert   error

### DISCUSSION THREAD

**Discussion by q1165600785** [sqlalchemy/sqlalchemy#9562] [Usage Questions]
def insert_model(self,conn,insert_dict_list):
        stmt=insert(antmod_one_one).values(
        antmod_one_one.c.model_name==bindparam("model_name"),
        antmod_one_one.c.antmod_name == bindparam("antmod_name"),
        antmod_one_one.c.dh == bindparam("dh"),
        antmod_one_one.c.rad == bindparam("rad"),
        antmod_one_one.c.L1_N == bindparam("L1_N"),
        antmod_one_one.c.L1_E == bindparam("L1_E"),
        antmod_one_one.c.L1_U == bindparam("L1_U"),
        antmod_one_one.c.L2_N == bindparam("L2_N"),
        antmod_one_one.c.L2_E == bindparam("L2_E"),
        antmod_one_one.c.L2_U == bindparam("L2_U"),
        antmod_one_one.c.remark == bindparam("remark"))
        result = conn.execute(stmt, insert_dict_list)
        return  result
if __name__ == '__main__':
    engine = create_engine(
        URL.create("mysql+pymysql",
                   username="root",
                   password="zhangyujie",
                   host="localhost",
                   port=3306,
                   database="antmod_database"), echo=True)
    with engine.connect() as conn:
        insert_dict_list=[
            {"model_name":"zz",
             "remark":"hh"
             },
            {
                "antmod_name":"yy",
                "dh":18
            },

        ]
        get_opr =model_table_opr()
        result=get_opr.insert_model(conn,insert_dict_list)
        print(result)

sqlalchemy.exc.ArgumentError: Only a single dictionary/tuple or list of dictionaries/tuples is accepted positionally.
how i can wirte code   correctly   please help me   thank you

**Comment by q1165600785** [comment]
stmt=insert(antmod_one_one).values(
        antmod_one_one.c.model_name==bindparam("model_name")).values(
        antmod_one_one.c.antmod_name == bindparam("antmod_name")).values(
        antmod_one_one.c.dh == bindparam("dh")).values(
        antmod_one_one.c.rad == bindparam("rad")).values(
        antmod_one_one.c.L1_N == bindparam("L1_N")).values(
        antmod_one_one.c.L1_E == bindparam("L1_E")).values(
        antmod_one_one.c.L1_U == bindparam("L1_U")).values(
        antmod_one_one.c.L2_N == bindparam("L2_N")).values(
        antmod_one_one.c.L2_E == bindparam("L2_E")).values(
        antmod_one_one.c.L2_U == bindparam("L2_U")).values(
        antmod_one_one.c.remark == bindparam("remark"))
error  AttributeError: Neither 'BinaryExpression' object nor 'Comparator' object has an attribute 'items' 

i am fish  i need help

**Reply by lelit** [reply]
Please, try to explain what the goal is, instead of just putting there a statement that raises an error.

If you want to insert values into a table, assuming `antmod_one_one` **is** a table, see [here](https://docs.sqlalchemy.org/en/20/tutorial/data_insert.html#tutorial-core-insert) for examples, and [here](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.bindparam) for `bindparam` ones.

**Reply by lelit** [reply]
`.values()` accepts keyword arguments, one per field, you are giving it a single boolean expression instead.

**Reply by q1165600785** [reply]
antmod_one_one=Table("antmod_one_one",metadata_obj,
    Column("id",BIGINT, primary_key=True),
    Column("model_name",VARCHAR(50)),
    Column("antmod_name",VARCHAR(50)),
    Column("dh",FLOAT(5,2)),
    Column("rad",FLOAT(5,2)),
    Column("L1_N",FLOAT(5,2)),
    Column("L1_E",FLOAT(5,2)),
    Column("L1_U",FLOAT(5,2)),
    Column("L2_N",FLOAT(5,2)),
    Column("L2_E",FLOAT(5,2)),
    Column("L2_U",FLOAT(5,2)),
    Column("add_time",DATETIME,default=datetime.datetime.now,onupdate=datetime.datetime.now),
    Column("source",TINYINT,default=1,onupdate=2),
    Column("remark",TEXT),
    )

i want  achieve batch insert and update  using  sqalchemy core   
but need care one point  that  ervery insert oject  has  different  field   
so  i  don't know  how  can  implement code

insert_dict_list=[
{"model_name":"zz",
"remark":"hh"
},
{
"antmod_name":"yy",
"dh":18
},

**Reply by q1165600785** [reply]
stmt=insert(antmod_one_one).values(model_name=bindparam("model_name"),
        antmod_name = bindparam("antmod_name"),
        dh = bindparam("dh"),
       rad = bindparam("rad"),
        L1_N = bindparam("L1_N"),
        L1_E = bindparam("L1_E"),
        L1_U = bindparam("L1_U"),
        L2_N = bindparam("L2_N"),
        L2_E = bindparam("L2_E"),
        L2_U = bindparam("L2_U"),
        remark = bindparam("remark"))

i correct  my  code  but  it still achieve  i want think

**Reply by q1165600785** [reply]
i correct my code but it still  not achieve i want think

**Reply by lelit** [reply]
Roughly something like the following:
```python
insert_dict_list=[
    {"model_name":"zz",
     "remark":"hh"},
    {"antmod_name":"yy",
     "dh":18},
]

for data in insert_dict_list:
    stmt = insert(antmod_one_one).values(**data)
    session.execute(stmt)
```

**Reply by q1165600785** [reply]
thank answer  i still  want to know  this dict type  it can not  achieve  bacth insert  or update?

**Reply by lelit** [reply]
I may be wrong, but I think that **batch** mode requires an uniform layout of the data, that is, all those dicts should have the same keys.
